cpp solution for usaco-623 (#3013)
cpinitiative/usaco-guide: solutions/usaco-623.mdx

---
id: usaco-623
source: USACO Gold 2016 February
title: Fenced In
author: Jeffrey Zhang
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_fencedin_gold_feb16.html)

## Solution

> The $n$ vertical and $m$ horizontal fences divide the large rectangular pasture into $(n+1)(m+1)$ grids. To help us better comprehend the problem, it is easier for us to visualized the input in terms of graphs: the grids themselves are the nodes, and the fence lengths of each 4 surrounding sides are the edge lengths that can be constructed to their neighboring nodes. For a cow to travel from any node to another, all we need is to construct a tree graph. And to ensure the edges picked are at a minimum, we just need to construct a minimum spanning tree, which can be solved by the [Kruskal's MST algorithm](https://usaco.guide/gold/mst#kruskals).

Reviewer: ```suggestion
The $n$ vertical and $m$ horizontal fences divide the large rectangular pasture into $(n+1)(m+1)$ grids. To better understand the problem, it is easier for us to visualize the input in terms of a graph: the grids themselves are the nodes, and the fences of each 4 surrounding sides are the edges that can be constructed. For a cow to travel from any node to another, all we need is to construct a tree. 

To ensure the edges picked are optimal, we construct a minimum spanning tree, which can be solved by [Kruskal's MST algorithm](https://usaco.guide/gold/mst#kruskals).
```

## Implementation

**Time Complexity:** $\mathcal{O}((N\text{max}(y_i))(\text{log}(\text{max}(y_i))+\text{log}(N)))$

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<int,int> pii;
#define ff first
#define ss second

//BeginCodeSnip{Disjoint-Set Union}

> pii find(pii target, vector<vector<pii>>& top){

Reviewer: i'd make a class for DSU, up to you though

Author: making an entirely new class might be a bit excessive...

Reviewer: fair enough

---
id: usaco-623
source: USACO Gold 2016 February
title: Fenced In
author: Jeffrey Zhang
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_fencedin_gold_feb16.html)

## Solution

The $n$ vertical and $m$ horizontal fences divide the large rectangular pasture into $(n+1)(m+1)$ grids. To better understand the problem, it is easier for us to visualize the input in terms of a graph: the grids themselves are the nodes, and the fences of each 4 surrounding sides are the edges that can be constructed. For a cow to travel from any node to another, all we need is to construct a tree. 

> To ensure the edges picked are optimal, we construct a minimum spanning tree, which can be solved by [Kruskal's MST algorithm](https://usaco.guide/gold/mst#kruskals).

Reviewer 2: ```suggestion
To ensure the edges picked are optimal, we construct a minimum spanning tree, which can be solved by [Kruskal's MST algorithm](/gold/mst#kruskals).
```

optional
